x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import datemath from '@elastic/datemath';
+import {
+  CommonProps,
+  EuiSuperDatePicker,
+  EuiSuperDatePickerProps
+} from '@elastic/eui';
+import { memoize } from 'lodash';
+import React from 'react';
+import { connect } from 'react-redux';
+import { RouteComponentProps, withRouter } from 'react-router-dom';
+import { updateTimePicker } from 'x-pack/plugins/apm/public/store/urlParams';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+interface Props extends RouteComponentProps {
+  dispatchUpdateTimePicker: typeof updateTimePicker;
+}
+
+interface DatePickerParams {
+  rangeFrom: string;
+  rangeTo: string;
+  refreshPaused: boolean;
+  refreshInterval: number;
+}
+
+// TODO this can be removed when EUI PR lands: https://github.com/elastic/eui/pull/1574
+declare module '@elastic/eui' {
+  interface OnTimeChangeProps {
+    start: string;
+    end: string;
+    isInvalid: boolean;
+    isQuickSelection: boolean;
+  }
+
+  interface OnRefreshChangeProps {
+    isPaused: boolean;
+    refreshInterval: number;
+  }
+
+  export type EuiSuperDatePickerProps = CommonProps & {
+    start?: string;
+    end?: string;
+    isPaused?: boolean;
+    refreshInterval?: number;
+    onTimeChange: (props: OnTimeChangeProps) => void;
+    onRefreshChange?: (props: OnRefreshChangeProps) => void;
+    showUpdateButton?: boolean;
+  };
+  export const EuiSuperDatePicker: React.SFC<EuiSuperDatePickerProps>;
+}
+
+const APM_DEFAULT_TIME_OPTIONS: DatePickerParams = {
+  rangeFrom: 'now-24h',
+  rangeTo: 'now',
+  refreshPaused: true,
+  refreshInterval: 0
+};
+
+class DatePickerComponent extends React.Component<Props> {
+  public state = {
+    from: '',
+    to: '',
+    isPaused: true,
+    refreshInterval: 0
+  };
+  public refreshTimeoutId = 0;
+
+  public getParamsFromSearch = memoize((search: string) => {
+    const query = toQuery(search);
+    const converted = {
+      ...query,
+      refreshPaused:
+        query.refreshPaused === 'true' || query.refreshPaused === true,
+      refreshInterval:
+        typeof query.refreshInterval === 'undefined'
+          ? undefined
+          : Number(query.refreshInterval)
+    };
+    return {
+      ...APM_DEFAULT_TIME_OPTIONS,
+      ...converted
+    } as DatePickerParams;
```

Comment:
TS doesn't understand that by defaulting to an object with all of the values defined, none of the four values in this interface can be undefined, so I've used `as` here to help TS understand.

Reply:
I don't think it's TS that can't handle this, rather that I have modelled it incorrectly in url_helpers. 
It can probably be done smarter, but this is at least a start:

```ts
interface APMQueryParamsRaw {
  transactionId?: string;
  traceId?: string;
  detailTab?: string;
  flyoutDetailTab?: string;
  waterfallItemId?: string;
  spanId?: string;
  page?: string;
  sortDirection?: string;
  sortField?: string;
  kuery?: string;
  rangeFrom?: string;
  rangeTo?: string;
  refreshPaused?: string;
  refreshInterval?: string;
}

interface APMQueryParamsBoth {
  transactionId?: string;
  traceId?: string;
  detailTab?: string;
  flyoutDetailTab?: string;
  waterfallItemId?: string;
  spanId?: string;
  page?: string | number;
  sortDirection?: string;
  sortField?: string;
  kuery?: string;
  rangeFrom?: string;
  rangeTo?: string;
  refreshPaused?: string | boolean;
  refreshInterval?: string | number;
}
```